shared/protos/compiler.proto

```diff
@@ -0,0 +1,93 @@
+syntax = "proto2";
+package buffers;
+
+import "game.proto";
+import "Settings.proto";
+
+message Empty {
+}
+
+message Resource {
+  optional string name = 1;
+  optional bool is_function = 2;
+  optional int32 arg_count_min = 3;
+  optional int32 arg_count_max = 4;
+  optional int32 overload_count = 5;
+  optional int32 is_type_name = 6;
+  optional int32 is_global = 7;
+  optional bool is_end = 8;
+  repeated string parameters = 9;
+}
+
+message CompileRequest {
+  enum CompileMode {
+    UNKNOWN = 0;
+    RUN = 1;
+    DEBUG = 2;
+    DESIGN = 3;
+    COMPILE = 4;
+    REBUILD = 5;
+  }
+
+  optional Game game = 1;
+  optional string name = 2;
+  optional CompileMode mode = 3;
+}
+
+message LogMessage {
+  enum Severity {
+    UNKNOWN = 0;
+    FINE = 1;
+    INFO = 2;
+    WARNING = 3;
+    ERROR = 4;
+  }
+  optional Severity severity = 1; ///< The severity and nature of the message.
+  optional string message = 2; ///< The message to display to the user in the output.
+}
+
+message ProgressMessage {
+  optional float progress = 1; ///< In range 1-100 (you can use 0-1, but take a stance, here)
+  optional string message = 2; ///< The message to display inside of or after the progress bar.
+}
+
+message CompileReply {
+  repeated LogMessage message = 1; ///< Compilation messages logged since the last update.
+  optional ProgressMessage progress = 2; ///< Current compilation progress.
+}
+
+message SyntaxError {
+  optional string message = 1; ///< The text associated with the error.
+  optional int32 line = 2; ///< The line number on which the error occurred.
+  optional int32 position = 3; ///< The column number in the line at which the error occurred.
+  optional int32 absolute_index = 4; ///< The zero-based position in the entire code at which the error occurred.
+}
+
+message SetDefinitionsRequest {
+  optional string code = 1;
+  optional string yaml = 2;
+}
+
+//NOTE: This request may change available keywords which should be reobtained with GetResources().
+message SetCurrentConfigRequest {
+  optional resources.Settings settings = 1; ///< The settings that you want to apply including API (e.g, "Direct3D9" for the graphics system).
+}
+
+message SyntaxCheckRequest {
```

Comment:
Where is the request message used if not in the RPC? I don't think there's a reason to separate request/response messages from the service definition. What's the point of building the request if there's nowhere to send it?

Reply:
The RPC stubs just call out to the compileEGMf functions. Also it doesn't matter, you said to use the Protocol Buffer interfaces directly for new stuff. That means in order to compile the compiler you'd need GRPC, but we don't want that. Emake could be built without server support but still do command line syntax checking.

Reply:
The new compile method doesn't accept a `CompileRequest`—it takes a `Project` proto and converts it to a `GameData` struct. To my knowledge, only the RPC service would accept a `CompileRequest`.

Reply:
I am sorry, I was talking about the new logging methods.
https://github.com/enigma-dev/enigma-dev/blob/4d62bf0170b47dad73ca5dc1e7eeea76ff7b033b/CommandLine/emake/Server.cpp#L140
https://github.com/enigma-dev/enigma-dev/blob/4d62bf0170b47dad73ca5dc1e7eeea76ff7b033b/CommandLine/emake/EnigmaCallbacks.cpp#L53

You are correct. Regardless, ideally, we'd just add the wrapper accepting SyntaxCheckRequest in compileEGMf, then have the old version call the new one. So again, we want to use these buffers in the compiler now, but don't want to make the compiler depend on GRPC by default.

Reply:
Okay, let's not get too carried away, here. Remember, you can't just pass the protos themselves across DLL bounds, anyway; you have to serialize them, first. I think we have emake just calling external methods accepting and returning protos all willy-nilly, and that works because emake and ENIGMA are built by the same compiler at the same time. But RGM and LGM cannot use that interface. They have to use an overload that accepts `byte[]` (or `char*`) and performs the serialization in the C++.

If that's where you want to use the request/response messages, then okay. I'd just as soon accept `(string, SomeSyntaxCheckMetadataProto)` (as bytes, not C++ types) and then use `optional string` and `SomeSyntaxCheckMetadataProto` as fields in the Request proto.

And before you ask, if you want to use threads with the non-RPC compiler API, you will want to use a ring buffer of string of serialized bytes, and a counting semaphore. Otherwise, just use a single static string.

Reply:
Ok, that all sounds about right and I had expected as much. I don't know why I don't think I actually considered the DLL bounds until you said that much though, but that would be extremely useful. The same should be done for libEGM, then people could redistribute it. The more third party tools that use our stuff, the more developers we will have that want to help maintain our stuff.